### ISSUE TITLE
write UUID fail on _check_schema_compatible

### ISSUE THREAD

**Issue by raphaelauv** [apache/iceberg-python#855]
### Apache Iceberg version

main (development)

### Please describe the bug 🐞

I can't write a UUID in an iceberg table

```python
from pyiceberg.catalog.rest import RestCatalog
from pyiceberg.schema import Schema
from pyiceberg.types import NestedField, UUIDType
import polars as pl
import uuid

catalog = RestCatalog(
    "default",
    **{
        "uri": "http://localhost:8181",
        "warehouse": "s3://test-bucket/",
        "s3.endpoint": "http://localhost:9020",
    },
)

catalog.create_namespace("default")
id_to_write = uuid.uuid4()

iceberg_schema = Schema(
    NestedField(1, "id", UUIDType(), required=True),
)
catalog.create_table(
    "default.aaa",
    schema=iceberg_schema,
)
df = pl.DataFrame({}).with_columns([pl.lit(id_to_write.bytes).alias("id")])

df = df.to_arrow()

df = df.cast(target_schema=iceberg_schema.as_arrow())

table = catalog.load_table("default.aaa")
table.append(df)

```

![image](https://github.com/apache/iceberg-python/assets/10202690/4da469af-bc3b-48d0-a916-fe7a89e73b93)

but if I comment the call to `_check_schema_compatible` then it write to the table

https://github.com/apache/iceberg-python/blob/a6cd0cf325b87b360077bad1d79262611ea64424/pyiceberg/table/__init__.py#L485

and I can read the data with trino

![Screenshot from 2024-06-25 13-43-09](https://github.com/apache/iceberg-python/assets/10202690/7d56357f-1053-4a1e-b83f-9c91f03f7de4)

**Comment by kevinjqliu** [comment]
thanks for reporting this issue! 

The `_check_schema_compatible` is currently more strict than it should be. In #829, the `_check_schema_compatible` check is relaxed. 
Would #829 fix your issue above?

**Comment by raphaelauv** [comment]
hey @kevinjqliu 
I tried your PR it do not fix the insert of UUID

**Comment by kevinjqliu** [comment]
I see, I also verified that `_check_schema_compatible` errors. 

Heres an example to repro:
```

def test_schema_uuid() -> None:
    import polars as pl

    iceberg_schema = Schema(
        NestedField(1, "id", UUIDType(), required=True),
    )

    id_to_write = uuid.uuid4()
    df = pl.DataFrame({}).with_columns([pl.lit(id_to_write.bytes).alias("id")])
    df = df.to_arrow()
    df = df.cast(target_schema=iceberg_schema.as_arrow())

    _check_schema_compatible(iceberg_schema, df.schema)
```

Looks like @Fokko opened an issue regarding UUID for arrow
https://github.com/apache/arrow/issues/15058

@Fokko can you chime in here on writing UUID data type?

**Comment by Fokko** [comment]
Thanks for pinging me here. So there is some progress on the Arrow side. There has been a vote to adopt the [UUID type](https://lists.apache.org/thread/94c7lby14j64k119tjzvpqsws24p2bbf), and it has been [added to the format](https://github.com/apache/arrow/pull/41299).

Thanks for the example code @kevinjqliu:

![image](https://github.com/apache/iceberg-python/assets/1134248/9ec4c599-a067-41e7-910e-728ab38c0c4a)

And I would say that they are equivalent. So if we know that the field in the Iceberg table is a UUID, just writing a `Fixed[16]` is okay and should pass the compatibility check.

**Comment by syun64** [comment]
Hi @raphaelauv thank you for raising this issue. https://github.com/apache/iceberg-python/pull/921 should fix this issue. Would you like to give it a try?

I've added a test including UUIDType to demonstrate that the fix will work

**Comment by kevinjqliu** [comment]
Fixed in #921

Specifically, tested here
https://github.com/apache/iceberg-python/pull/921/files#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128R2535

https://github.com/apache/iceberg-python/pull/921/files#diff-7f3dd1244d08ce27c003cd091da10aa049f7bb0c7d5397acb4ec69767036accdR1044
